dianatanase/assignment2/Q1.java

```diff
@@ -0,0 +1,159 @@
+import java.util.ArrayList;
+
+// Represent the Tree with Nodes that store data of generic type T.
+// Each node as a key, I assumed keys are positive integers and unique for each node.
+// Each node has a leaf and right child (could be null = leaf).
+// Store the parent of each node as well.
+class Node<T>{
+	
+	T data = null;
+	Node<T> left = null;
+	Node<T> right = null;
+	Node<T> parent = null;
+	int key = -1;
+	
+	Node(Node<T> parent, int key, T data, Node<T> left, Node<T> right){
+		this.data = data;
+		this.parent = parent;
+		this.left = left;
+		this.right = right;
+		this.key = key;
+	}
+	
+	// Update left node, right node and the data.
+	public void leftNode(Node<T> left){
+		this.left = left;
+	}
+	
+	public void rightNode(Node<T> right){
+		this.right = right;
+	}
+	
+	public void updateData(T data){
+		this.data = data;
+	}
+	
+	public T getData(){
+		return data;
+	}
+	
+	// Update parent in case we want to move a node. 
+	public void updateParent(Node<T> parent){
```

Comment:
It can be dangerous to both include the ability to set child nodes AND change the parent. The node can get out of sync very easily if you forget to do both. Why not update the value of the children's parent as part of leftNode() & rightNode() ?